Mathlib/Order/DirectedInverseSystem.lean

```diff
@@ -0,0 +1,315 @@
+
+Now our task reduces to the recursive construction of a *natural* family of bijections for each `i`.
+We can prove that a natural family over all `l ≤ i` (`Iic i`) extends to a natural family over
+`Iic i⁺` (where `i⁺ = succ i`), but at a limit node, recursion stops working: we have natural
+families over all `Iic j` for each `j < i`, but we need to know that they glue together to form a
+natural family over all `l < i` (`Iio i`). This intricacy did not occur to the author when he
+thought he had a proof and set out to formalize it. Fortunately he was able to figure out an
+additional `compat` condition (compatibility with the bijections `F i⁺ ≃ F i × X i` in the `X`
+component) that guarantees uniqueness (`unique_pEquivOn`) and hence gluability (well-definedness):
+see `pEquivOnGlue`. Instead of just a family of natural families, we actually construct a family of
+the stronger `PEquivOn`s that bundles the `compat` condition, in order for the inductive argument
+to work.
+
+It is possible to circumvent the introduction of the `compat` condition using Zorn's lemma;
+if there is a chain of natural families (i.e. for any two families in the chain, one is an
+extension of the other) over lowersets (which are all of the form `Iic`, `Iio`, or `univ`),
+we can clearly take the union to get a natural family that extends them all. If a maximal
+natural family has domain `Iic i` or `Iio i` (`i` a limit), we already know how to extend it
+one step further to `Iic i⁺` or `Iic i` respectively, so it must be the case that the domain
+is everything. However, the author chose the `compat` approach in the end because it constructs
+the distinguished bijection that is compatible with the projections to all `X i`.
+
+-/
+
+open Order Set
+
+variable {ι : Type*} [Preorder ι] {F X : ι → Type*}
+
+variable (F) in
+/-- A directed system is a functor from a category (directed poset) to another category. -/
+class DirectedSystem (f : ∀ ⦃i j⦄, i ≤ j → F i → F j) : Prop where
+  map_self ⦃i⦄ (x : F i) : f le_rfl x = x
+  map_map ⦃k j i⦄ (hij : i ≤ j) (hjk : j ≤ k) (x : F i) : f hjk (f hij x) = f (hij.trans hjk) x
+
+variable (f : ∀ ⦃i j : ι⦄, i ≤ j → F j → F i) ⦃i j : ι⦄ (h : i ≤ j)
+
+/-- A inverse system indexed by a preorder is a contravariant functor from the preorder
+to another category. It is dual to `DirectedSystem`. -/
+class InverseSystem : Prop where
+  map_self ⦃i⦄ (x : F i) : f le_rfl x = x
+  map_map ⦃k j i⦄ (hkj : k ≤ j) (hji : j ≤ i) (x : F i) : f hkj (f hji x) = f (hkj.trans hji) x
+
+namespace InverseSystem
+
+section proj
+
+/-- The inverse limit of an inverse system of types. -/
+def limit (i : ι) : Set (∀ l : Iio i, F l) :=
+  {F | ∀ ⦃j k⦄ (h : j.1 ≤ k.1), f h (F k) = F j}
+
+/-- For a family of types `X` indexed by an preorder `ι` and an element `i : ι`,
+`piLT X i` is the product of all the types indexed by elements below `i`. -/
+abbrev piLT (X : ι → Type*) (i : ι) := ∀ l : Iio i, X l
```

Comment:
Maybe nitpicky, but why not use `Π` in the definition?
```suggestion
abbrev piLT (X : ι → Type*) (i : ι) := Π l : Iio i, X l
```

Reply:
I think the ship has sailed and we are not using `Π` anymore

Reply:
Yeah, I think mathport translated all Π to ∀. (The pretty printer does neither; it prints `(i : ι) → f i`.) [DFinsupp](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/DFinsupp/Basic.html#DFinsupp) still uses a Π₀ notation though. I intend to use ∀ but may inadvertently put Π occasionally.